Transformation and URDF Setup Guides (#117)
ros-planning/navigation.ros.org: setup_guides/urdf/setup_urdf.rst

.. _urdf_handson:

Setting Up The URDF
###################

> For this guide, we will be creating the URDF for a simple differential drive robot to give you hands-on experience on working with URDF. We will also setup the robot state publisher and visualize our model in RVIZ. Lastly, we will be adding some kinematic properties to our robot URDF to prepare it for simulation purposes. 

Reviewer: Add a sentence about why we're doing this (e.g.  this is needed to represent all the sensor, hardware, and robot transforms of your robot for use in navigation)

URDF and the Robot State Publisher
==================================

> As discussed in the previous section, one of the requirements for Navigation2 is the  ``base_link`` => ``laser_link`` transformations. This transformation tree can range from a simple tree with only one link from the  ``base_link`` to ``laser_link`` or a tree comprised of multiple sensors located in different locations, each having their own coordinate frame. Creating multiple publishers to handle all of these coordinate frame transformations may become tedious. Therefore, we will be making use of the Robot State Publisher package to publish our transforms. 

Reviewer: ```suggestion
As discussed in the previous tutorial, one of the requirements for Navigation2 is the transformation from  ``base_link`` to the various sensors and reference frames. This transformation tree can range from a simple tree with only one link from the  ``base_link`` to ``laser_link`` or a tree comprised of multiple sensors located in different locations, each having their own coordinate frame. Creating multiple publishers to handle all of these coordinate frame transformations may become tedious. Therefore, we will be making use of the Robot State Publisher package to publish our transforms. 
```

Another major feature of URDF is that it also supports Xacro (XML Macros) to help you create a shorter and readable XML to help in defining complex robots. We can use these macros to eliminate the need for repeating blocks of XML in our URDF. Xacro is also useful in defining configuration constants which can be reused throughout the URDF.

.. seealso::
  If you want to learn more about the URDF and the Robot State Publisher, we encourage you to have a look at the official `URDF Documentation <http://wiki.ros.org/urdf>`__ and `Robot State Publisher Documentation <http://wiki.ros.org/robot_state_publisher>`__  

Setting Up the Environment
==========================

> .. warning:: This tutorial was tested and written with ROS2 Foxy. However, we always recommend to use the latest version of ROS2 if possible.

Reviewer: Remove this `warning` block and replace any instances of `foxy` with `<ros2-distro>`. Its what we do in all other docs to keep them independent of version to age better. e.x. `ros-<ros2-distro>-xacro`

In this guide, we are assuming that you are already familiar with ROS2 and how to setup your development environment, so we'll breeze through the steps in this section.

> .. seealso:: If you are new to ROS2 or do not have a working environment yet, then please take some time to properly setup your machine using the resources in the official `ROS2 Installation Documentation <https://index.ros.org/doc/ros2/Installation/>`__

Reviewer: Remove this block, we can assume they're going through these pages in order and this appeared in the last one

Reviewer: (and the prior line makes it clear the assumption is that's already done)

Let's begin by installing some additional ROS2 packages that we will be using during this tutorial.

.. code-block:: shell

  sudo apt install ros-foxy-joint-state-publisher-gui
  sudo apt install ros-foxy-xacro
 

> Next, create a directory for your project, initialize a ROS2 workspace and give your robot a name. For ours, we'll be calling it ``sam-bot``.

Reviewer: Use only underscores, `sam_bot`. Please update here and anywhere else `-` exists.

Reviewer: call the new package `sam_bot_description` thats the standard ROS package name for URDF files

Now that we have our project workspace set up, let's dive straight into writing the URDF. Below is an image of the robot we will be trying to build.

.. image:: images/base-bot_1.png
   :width: 49%
.. image:: images/base-bot_2.png
   :width: 49%

|

 To get started, create a file named ``sam-bot_description.urdf`` under ``src/description`` and input the following as the initial contents of the file. 

.. code-block:: xml
  :linenos:

  <?xml version="1.0"?>
  <robot name="sam-bot" xmlns:xacro="http://ros.org/wiki/xacro">

  </robot>

.. note:: The following code snippets should be placed within the ``<robot>`` tags. We suggest to add them in the same order as introduced in this tutorial. We have also included some line numbers to give you a rough idea on where to input the code. This may differ from the actual file you are writing depending on your usage of whitespaces. Also note that the line numbers assume that you are putting in code as they appear in this guide.

> Next, let us define some constants that will be used for different sections of our URDF. Note that we will be using XAcro properties to achieve this. 

Reviewer: Please have a short couple of sentences explaining what these are. I think its pretty clear from the code, but its always good to be explicit on each of the values.

.. code-block:: xml
  :lineno-start: 4

    <!-- Define robot constants -->
    <xacro:property name="base_width" value="0.31"/>
    <xacro:property name="base_length" value="0.42"/>
    <xacro:property name="base_height" value="0.18"/>

    <xacro:property name="wheel_radius" value="0.10"/>
    <xacro:property name="wheel_width" value="0.04"/>
    <xacro:property name="wheel_ygap" value="0.025"/>
    <xacro:property name="wheel_zoff" value="0.05"/>
    <xacro:property name="wheel_xoff" value="0.12"/>

    <xacro:property name="caster_xoff" value="0.14"/>

Let us then define our ``base_link`` - this link will be a large box and will act as the main chassis of our robot. In URDF, a ``link`` element describes a rigid part or component of our robot. The robot state publisher then utilizes these definitions to determine coordinate frames for each link and publish the transformations between them. 

> We will also be defining some of the link's visual properties which can be used by tools such as Gazebo and Rviz to show us a 3D model of our robot. Amongst these properties are `<geometry>` which describes the link's shape and `<material>` which describes it's color.

Reviewer: Same here, please go through the code block and explain what's happening a little. I think things to mention
- We're accessing the xacro variables from before to set our robot base size using their names as parameters in `${}`.
- The color will be green-blue because of the color choice 
- Its under the visual tag because this is just visualization parameters, which don't impact collision or physical properties.

.. code-block:: xml
  :lineno-start: 17

    <!-- Robot Base -->
    <link name="base_link">
      <visual>
        <geometry>
          <box size="${base_length} ${base_width} ${base_height}"/>
        </geometry>
        <material name="Cyan">
          <color rgba="0 1.0 1.0 1.0"/>
        </material>
      </visual>
    </link>

Next, let us define a ``base_footprint`` link. The ``base_footprint`` link is a virtual (non-physical) link which has no dimensions or collision areas. Its primary purpose is to enable various packages determine the center of a robot projected to the ground. For example, Navigation2 uses this link to determine the center of a circular footprint used in its obstacle avoidance algorithms. Again, we set this link with no dimensions and to which position the robot's center is in when it is projected to the ground plane.

> After defining our base_link, we then add a joint to connect it to ``base_link``. In URDF, a ``joint`` element describes the kinematic and dynamic properties between coordinate frames. For this case, we will be defining a ``fixed`` joint with the appropriate offsets to place our ``base_footprint`` link in the proper location based on the description above.

Reviewer: Please explain your math on the joint Z offset.

.. code-block:: xml

  <exec_depend>joint_state_publisher</exec_depend>
  <exec_depend>joint_state_publisher_gui</exec_depend>
  <exec_depend>robot_state_publisher</exec_depend>
  <exec_depend>rviz</exec_depend>
  <exec_depend>xacro</exec_depend>

Next, let us create our launch file. Launch files are used by ROS2 to bring up the necessary nodes for our package. From the root of the project, create a directory named ``launch`` and a ``display.launch.py`` file within it. The launch file below launches a robot publisher node in ROS2 that uses our URDF to publish the transforms for our robot. In addition, the launch file also automatically launches RVIZ so we can visualize our robot as defined by the URDF. Copy and paste the snippet below into your ``display.launch.py`` file. 

.. code-block:: python

  import launch
  from launch.substitutions import Command, LaunchConfiguration
  import launch_ros
  import os

  def generate_launch_description():
      pkg_share = launch_ros.substitutions.FindPackageShare(package='sam-bot').find('sam-bot')

>       default_model_path = os.path.join(pkg_share, 'src/description/sam-bot_description.urdf')

Reviewer: I didn't notice any changes to the cmakelists file, does this install the URDF to be able to do this?

Author: Oh I missed this one., already added snippet in f1acdef

          Value: /robot_description
        Enabled: true
        Name: RobotModel
        Visual Enabled: true
      - Class: rviz_default_plugins/TF
        Enabled: true
        Name: TF
        Marker Scale: 0.3
        Show Arrows: true
        Show Axes: true
        Show Names: true
    Enabled: true
    Global Options:
      Background Color: 48; 48; 48
      Fixed Frame: base_link
      Frame Rate: 30
    Name: root
    Tools:
      - Class: rviz_default_plugins/Interact
        Hide Inactive Objects: true
      - Class: rviz_default_plugins/MoveCamera
      - Class: rviz_default_plugins/Select
      - Class: rviz_default_plugins/FocusCamera
      - Class: rviz_default_plugins/Measure
        Line color: 128; 128; 0
    Transformation:
      Current:
        Class: rviz_default_plugins/TF
    Value: true
    Views:
      Current:
        Class: rviz_default_plugins/Orbit
        Name: Current View
        Target Frame: <Fixed Frame>
        Value: Orbit (rviz)
      Saved: ~

We are now ready to build our project using colcon. Navigate to the project root and execute the following commands.

.. code-block:: shell

  colcon build
  . install/setup.bash

> .. note:: If you do not have colcon set up yet, you may install it through the command ``sudo apt install python3-colcon-common-extensions`` or through your preferred package manager.

Reviewer: You can remove this, this would be covered by a ros 2 setup

After a successful build, execute the following commands to install the ROS2 package and launch our project.

.. code-block:: shell

>   ros2 launch launch/display.launch.py

Reviewer: I think it would be good to use the full syntax, if it works for this demo, like `ros2 launch <pkg-name> <launch-file>` though that may require cmakelists updates to install them

.. code-block:: xml
  :lineno-start: 113

      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${-(wheel_radius+wheel_zoff-(base_height/2))}"/>
        </geometry>
      </collision>

      <xacro:sphere_inertia m="0.5" r="${-(wheel_radius+wheel_zoff-(base_height/2))}"/>

.. note:: We did not add any inertial or collision properties to our ``base_footprint`` link since this is a virtual and non-physical link.

Build your project and then launch RViz using the same commands in the previous section.

.. code-block:: shell

  colcon build
  . install/setup.bash
  ros2 launch launch/display.launch.py

You can verify whether you have properly set up the collision areas by enabling ``Collision Enabled`` under ``RobotModel`` on the left pane (it may be easier to see if you also turn off ``Visual Enabled``). For this tutorial we defined a collision area which is similar to our visual properties. Note that this may not always be the case since you may opt for simpler collision areas based on how your robot looks.

.. image:: images/base-bot_5.png

> For now, we will have to stop here since we will need to set up a lot more components to actually start simulating our robot in Gazebo. We will be coming back to this project during the course of these setup guides, and we will eventually see our robot move in a virtual environment once we get to the simulation sections.

Reviewer: Add "The only major thing missing from this to work in gazebo is the simulation plugins required to mimic your robot controllers. We will introduce those and add them to this UDRF in the appropriate section."

.. _urdf_handson:

Setting Up The URDF
###################

For this guide, we will be creating the URDF for a simple differential drive robot to give you hands-on experience on working with URDF. We will also setup the robot state publisher and visualize our model in RVIZ. Lastly, we will be adding some kinematic properties to our robot URDF to prepare it for simulation purposes. These steps are necessary to represent all the sensor, hardware, and robot transforms of your robot for use in navigation.

> 

Reviewer: Please place the image from slide 20 here https://mirror.umd.edu/roswiki/attachments/Events(2f)PR2BetaTraining/tf.pdf. Its nice to show a visual of what a "real" one of these would look like as example to start.

  <robot name="sam_bot" xmlns:xacro="http://ros.org/wiki/xacro">

  </robot>

.. note:: The following code snippets should be placed within the ``<robot>`` tags. We suggest to add them in the same order as introduced in this tutorial. We have also included some line numbers to give you a rough idea on where to input the code. This may differ from the actual file you are writing depending on your usage of whitespaces. Also note that the line numbers assume that you are putting in code as they appear in this guide.

Next, let us define some constants using XAcro properties that will be reused throughout the URDF.

.. code-block:: xml
  :lineno-start: 4

    <!-- Define robot constants -->
    <xacro:property name="base_width" value="0.31"/>
    <xacro:property name="base_length" value="0.42"/>
    <xacro:property name="base_height" value="0.18"/>

    <xacro:property name="wheel_radius" value="0.10"/>
    <xacro:property name="wheel_width" value="0.04"/>
    <xacro:property name="wheel_ygap" value="0.025"/>
    <xacro:property name="wheel_zoff" value="0.05"/>
    <xacro:property name="wheel_xoff" value="0.12"/>

    <xacro:property name="caster_xoff" value="0.14"/>

> Here is a brief discussion on what these properties will represent in our urdf. The ``base_*`` properties all define the size of hte robot's main chassis. The ``wheel_radius`` and ``wheel_width`` define the shape of the robot's two back wheels. The `wheel_ygap` adjusts the gap between the wheel and the chassis along the y-axis whilst ``wheel_zoff`` and ``wheel_xoff`` position the back wheels along the z-axis and x-axis appropriately. Lastly, the ``caster_xoff`` positions the front caster wheel along the x-axis.

Reviewer: ```suggestion
Here is a brief discussion on what these properties will represent in our urdf. The ``base_*`` properties all define the size of hte robot's main chassis. The ``wheel_radius`` and ``wheel_width`` define the shape of the robot's two back wheels. The ``wheel_ygap`` adjusts the gap between the wheel and the chassis along the y-axis whilst ``wheel_zoff`` and ``wheel_xoff`` position the back wheels along the z-axis and x-axis appropriately. Lastly, the ``caster_xoff`` positions the front caster wheel along the x-axis.
```

Let us then define our ``base_link`` - this link will be a large box and will act as the main chassis of our robot. In URDF, a ``link`` element describes a rigid part or component of our robot. The robot state publisher then utilizes these definitions to determine coordinate frames for each link and publish the transformations between them. 

> We will also be defining some of the link's visual properties which can be used by tools such as Gazebo and Rviz to show us a 3D model of our robot. Amongst these properties are `<geometry>` which describes the link's shape and `<material>` which describes it's color.

Reviewer: ```suggestion
We will also be defining some of the link's visual properties which can be used by tools such as Gazebo and Rviz to show us a 3D model of our robot. Amongst these properties are ``<geometry>`` which describes the link's shape and ``<material>`` which describes it's color.
```
